### PR TITLE
#618 filters.custom can specify regex

### LINKED ISSUE #618: [Feature Request] Filtering by glob
Feature is present in https://github.com/ms-jpq/chadtree (GIF there showing it in action).

Suppose you have a typical project file tree like:
```
.
├── Dockerfile
├── README.md
├── api
│   ├── bulk
│   │   └── endpoint.go
│   └── graph
│       ├── endpoint.go
│       ├── generated
│       │   └── generated.go
│       ├── model
│       │   ├── models.go
│       │   ├── models_gen.go
│       │   └── namespaces
│       │       └── namespaces.go
│       ├── resolver
│       │   ├── common.resolvers.go
│       │   ├── filters.resolvers.go
│       │   ├── resolver.go
│       │   └── schema
│       │       ├── common.graphql
│       │       └── filters.graphql
│       └── signing
│           ├── signing.go
│           └── signing_test.go
├── config
│   └── config.go
├── config.yaml
├── db
│   └── migrations
│       └── 0000_average_spend.up.sql
├── docker-compose.yaml
├── go.mod
├── go.sum
├── gqlgen.yml
├── kube
│   └── dev.yaml
├── main.go
├── services
│   ├── ingestion.go
│   └── services.go
├── storage
│   ├── dataset.go
│   ├── filters.go
│   ├── filters_test.go
│   └── postgres.go
├── terraform
│   ├── dev-postgres.tf
│   └── main.tf
└── util
    ├── log
    │   └── log.go
    └── types
        └── decimal.go
```

and I want to see the tree of all the graphql files only via applying a filter `*.graphql`, I'd expect the tree to appear as

```
.
└── api
    └── graph
        └── resolver
            └── schema
                ├── common.graphql
                └── filters.graphql
```

Regardless if this is implemented or not, awesome project @kyazdani42, you're awesome!

### PR DESCRIPTION
**Pull request by alex-courtis**
closes #618

### CODE REVIEW

File: lua/nvim-tree/explorer/filters.lua [hunk 1]
```diff
@@ -35,8 +35,14 @@ function M.should_ignore(path)
...
 
   local relpath = utils.path_relative(path, vim.loop.cwd())
-  if M.ignore_list[relpath] == true or M.ignore_list[basename] == true then
-    return true
+  for pat, _ in pairs(M.ignore_list) do
+    if pat == relpath or pat == basename then
+      return true
+    end
+    pat = "^" .. pat .. "$"
+    if relpath:match(pat) or basename:match(pat) then
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/filters.lua:43]: We could use vim.regex if you prefer.

Comment (kyazdani42) [reply lua/nvim-tree/explorer/filters.lua:43]: i think it would be better indeed, lua regexes are not very common and harder to use than usual regexes :)

**Comment by kyazdani42** [comment]
also this does not close #618 , the feature request asks for the live-filter feature which i'm currently implementing :)

File: lua/nvim-tree/explorer/filters.lua (hunk 1 above)
Comment (alex-courtis) [reply lua/nvim-tree/explorer/filters.lua:43]: Sensible. Will change on weekend and update the doc.

**Review by kyazdani42 — APPROVED** [review]

**Comment by mvllow** [comment]
This causes unexpected behaviour when trying to filter a specific file, eg. `custom = { '.git' }` now hides `.github` and `.gitignore`. Using `.git$` works but not sure if that's the most obvious solution to most (I had to look it up). Seeing how this is already merged I don't expect any changes but maybe a note/example in the readme would be helpful 😌

**Comment by alex-courtis** [comment]
> If there is an appropriate page in the wiki for snippets or examples I'd be happy to edit that as well

I very briefly documented the option in the [help](https://github.com/kyazdani42/nvim-tree.lua/blob/master/doc/nvim-tree-lua.txt#L410) however it could be clearer. A PR to update that would be great!

**Pull request merged**